implants/lib/eldritch/src/pivot/arp_scan_impl.rs

```diff
@@ -1,5 +1,199 @@
+            None => continue,
+        };
+        let mut arp_packet = [0u8; 28];
+        let mut arp = MutableArpPacket::new(&mut arp_packet).unwrap();
+        arp.set_hardware_type(pnet::packet::arp::ArpHardwareType(1));
+        arp.set_protocol_type(pnet::packet::ethernet::EtherType(0x0800));
+        arp.set_hw_addr_len(6);
+        arp.set_proto_addr_len(4);
+        arp.set_operation(ArpOperations::Request);
+        arp.set_sender_hw_addr(mac);
+        arp.set_sender_proto_addr(ip_to_use.ip());
+        arp.set_target_hw_addr(MacAddr::zero());
+        arp.set_target_proto_addr(ip);
+        let mut eth_packet = [0u8; 60];
+        let mut eth = MutableEthernetPacket::new(&mut eth_packet).unwrap();
+        eth.set_destination(MacAddr::broadcast());
+        eth.set_source(mac);
+        eth.set_ethertype(EtherType(0x0806));
+        eth.set_payload(arp.packet());
+        let tx_res = tx.send_to(eth.packet(), None).unwrap();
+        if let Err(err) = tx_res {
+            println!("Failed to tx on dev {}: {}", interface.name, err);
+            return;
+        }
+        let now = SystemTime::now();
+        loop {
+            if now.elapsed().unwrap() > Duration::from_secs(5) {
+                break;
+            }
+            match rx.next() {
+                Ok(packet) => {
+                    let eth = match EthernetPacket::new(packet) {
+                        Some(eth) => eth,
+                        None => continue,
+                    };
+                    let arp = match ArpPacket::new(eth.payload()) {
+                        Some(arp) => arp,
+                        None => continue,
+                    };
+                    if arp.get_operation() != ArpOperations::Reply {
+                        continue;
+                    }
+                    let source_ip = arp.get_sender_proto_addr();
+                    let source_mac = arp.get_sender_hw_addr();
+                    if source_ip == ip {
+                        *data.lock().unwrap().get_mut(&ip).unwrap() = Some(ArpResponse {
+                            _source_ip: source_ip,
+                            source_mac,
+                            interface: interface.name.clone(),
+                        });
+                        break;
+                    }
+                }
+                Err(e) => {
+                    if e.kind() == std::io::ErrorKind::TimedOut {
+                        continue;
+                    }
+                    println!("Error receiving packet: {}", e);
+                    return;
+                }
+            }
+        }
+    }
+}
+
+pub fn arp_scan(starlark_heap: &Heap, target_cidrs: Vec<String>) -> Result<Vec<Dict>> {
+    let target_cidrs = target_cidrs
+        .iter()
+        .map(|cidr| {
+            let (addr, prefix) = cidr.split_at(cidr.find('/').unwrap());
+            let addr = match Ipv4Addr::from_str(addr) {
+                Ok(addr) => addr,
+                Err(_) => {
```

Comment:
I have a medium preference towards returning the original error string at the end of the anyhow error but I’ll defer to you if that makes sense Eg if no other error can return may not make sense.